File: src/main/java/io/github/roguesdev/hoard/DefaultReactiveStreamDepositor.java
package io.github.roguesdev.hoard;

import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

class DefaultReactiveStreamDepositor<T> implements ReactiveStreamDepositor<T> {

  private final Depositor<T> depositor;

  DefaultReactiveStreamDepositor(Depositor<T> depositor) {
    this.depositor = depositor;
  }

  @Override public Publisher<Void> store(final T value) {
    return new Publisher<Void>() {
      @Override public void subscribe(final Subscriber<? super Void> s) {
        s.onSubscribe(new Subscription() {

          volatile boolean canceled = false;

          @Override public void request(long n) {
            try {
              depositor.store(value);

              if (canceled) return;
              s.onComplete();
            } catch (Exception e) {
              if (canceled) return;
              s.onError(e);
            }
          }

          @Override public void cancel() {
            canceled = true;
          }
        });
      }
    };
  }

  @Override public Publisher<T> retrieve() {
    return new Publisher<T>() {
      @Override public void subscribe(final Subscriber<? super T> s) {
        s.onSubscribe(new Subscription() {

          volatile boolean canceled = false;

          @Override public void request(long n) {
            try {
              T value = depositor.retrieve();
              if (canceled) return;
              if (value != null) {
                s.onNext(value);
              }
              s.onComplete();
            } catch (Exception e) {
              if (canceled) return;
              s.onError(e);
            }
          }

          @Override public void cancel() {
            canceled = true;
          }
        });
      }
    };
  }

  @Override public Publisher<Void> delete() {
    return new Publisher<Void>() {
      @Override public void subscribe(final Subscriber<? super Void> s) {
        s.onSubscribe(new Subscription() {

          volatile boolean canceled = false;

          @Override public void request(long n) {
            try {
              depositor.delete();
              if (canceled) return;
              s.onComplete();
            } catch (Exception e) {
              if (canceled) return;
              s.onError(e);
            }
          }

          @Override public void cancel() {
            canceled = true;
          }
        });
      }
    };
  }

  @Override public Publisher<Boolean> exists() {
    return new Publisher<Boolean>() {
      @Override public void subscribe(final Subscriber<? super Boolean> s) {
        s.onSubscribe(new Subscription() {

          volatile boolean canceled = false;

          @Override public void request(long n) {
            try {
              boolean exists = depositor.exists();

              if (canceled) return;
              s.onNext(exists);
              s.onComplete();
            } catch (Exception e) {
              if (canceled) return;
              s.onError(e);
            }
          }

          @Override public void cancel() {

          }
        });
      }
    };
  }
}

File: src/main/kotlin/io/github/roguesdev/hoard/DefaultReactiveStreamDepositor.kt
package io.github.roguesdev.hoard

import org.reactivestreams.Publisher
import org.reactivestreams.Subscription

internal class DefaultReactiveStreamDepositor<T>(private val depositor: Depositor<T>) : ReactiveStreamDepositor<T> {

  override fun store(value: T): Publisher<Void> {
    return Publisher { s ->
      s.onSubscribe(object : Subscription {

        @Volatile
        var canceled = false

        override fun request(n: Long) {
          try {
            depositor.store(value)

            if (canceled) return
            s.onComplete()
          } catch (e: Exception) {
            if (canceled) return
            s.onError(e)
          }
        }

        override fun cancel() {
          canceled = true
        }
      })
    }
  }

  override fun retrieve(): Publisher<T> {
    return Publisher { s ->
      s.onSubscribe(object : Subscription {

        @Volatile
        var canceled = false

        override fun request(n: Long) {
          try {
            val value = depositor.retrieve()
            if (canceled) return
            if (value != null) {
              s.onNext(value)
            }
            s.onComplete()
          } catch (e: Exception) {
            if (canceled) return
            s.onError(e)
          }
        }

        override fun cancel() {
          canceled = true
        }
      })
    }
  }

  override fun delete(): Publisher<Void> {
    return Publisher { s ->
      s.onSubscribe(object : Subscription {

        @Volatile var canceled = false

        override fun request(n: Long) {
          try {
            depositor.delete()
            if (canceled) return
            s.onComplete()
          } catch (e: Exception) {
            if (canceled) return
            s.onError(e)
          }
        }

        override fun cancel() {
          canceled = true
        }
      })
    }
  }

  override fun exists(): Publisher<Boolean> {
    return Publisher { s ->
      s.onSubscribe(object : Subscription {

        @Volatile var canceled = false

        override fun request(n: Long) {
          try {
            val exists = depositor.exists()

            if (canceled) return
            s.onNext(exists)
            s.onComplete()
          } catch (e: Exception) {
            if (canceled) return
            s.onError(e)
          }
        }

        override fun cancel() = Unit
      })
    }
  }
}
